File: packages/ik_rules/lib/src/bounties/complete.dart
import 'dart:math' as math;

import '../js_compat.dart';
import '../production/recipes.dart';
import '../save/generated/save_models.dart';
import 'progress.dart';
import 'rotation.dart';
import 'types.dart';

/// Either the save with the turn-in staged, or the reason it was refused.
class BountyTurnInCheck {
  const BountyTurnInCheck.ok(this.save) : reason = null;

  const BountyTurnInCheck.failed(this.reason) : save = null;

  bool get ok => reason == null;
  final PlayerSave? save;
  final String? reason;

  Map<String, Object?> toJson() => ok
      ? <String, Object?>{'ok': true, 'save': save!.toJson()}
      : <String, Object?>{'ok': false, 'reason': reason};
}

BountyTurnInCheck _consumeInventoryItems(PlayerSave save, String itemId, num amount) {
  if (inventoryCount(save, itemId) < amount) {
    return const BountyTurnInCheck.failed('Not enough items to deliver.');
  }
  var remaining = amount;
  final inventory = <InventoryStack>[];
  for (final stack in save.inventory) {
    if (stack.itemId != itemId || remaining <= 0 || isNotBlank(stack.enchantmentId)) {
      if (stack.quantity > 0) inventory.add(stack);
      continue;
    }
    final take = math.min(stack.quantity, remaining);
    remaining -= take;
    final left = stack.quantity - take;
    if (left > 0) inventory.add(stack.copyWith(quantity: left));
  }
  if (remaining > 0) return const BountyTurnInCheck.failed('Not enough items to deliver.');
  return BountyTurnInCheck.ok(save.copyWith(inventory: inventory));
}

/// Everything the turn-in decides locally: the board is current, the objective
/// is met, and any delivered items have left the bag. The returned save is what
/// the reward applies to once a backend accepts the claim.
BountyTurnInCheck prepareBountyTurnIn(PlayerSave save, BountyDefinition bounty, num nowMs) {
  final next = syncBountyHour(save, nowMs);
  final board = hourlyBountyBoard(nowMs);
  if (board.hourKey != next.bountyHourKey) {
    return const BountyTurnInCheck.failed('This bounty board has rotated.');
  }
  if (!board.bounties.any((row) => row.id == bounty.id)) {
    return const BountyTurnInCheck.failed('That bounty is not on the current board.');
  }
  if (next.bountyClaimedIds.contains(bounty.id)) {
    return const BountyTurnInCheck.failed('You already claimed this bounty.');
  }
  if (!isBountyReadyToClaim(next, bounty, nowMs)) {
    return const BountyTurnInCheck.failed('Finish the bounty objective first.');
  }
  if (bounty.kind == 'gather_deliver') {
    return _consumeInventoryItems(next, bounty.targetId, bounty.amount);
  }
  return BountyTurnInCheck.ok(next);
}

class BountyRewardResult {
  const BountyRewardResult({required this.save, required this.goldGained});

  final PlayerSave save;
  final num goldGained;

  Map<String, Object?> toJson() => <String, Object?>{
    'save': save.toJson(),
    'goldGained': goldGained,
  };
}

/// Pays out a bounty a backend has accepted. Everyone eligible earns the base
/// reward; only the claim the server recorded first adds the bonus.
BountyRewardResult applyBountyReward(
  PlayerSave save,
  BountyDefinition bounty,
  bool firstCompleter,
) {
  final goldGained = bounty.rewardGold + (firstCompleter ? bounty.firstPlaceBonusGold : 0);
  return BountyRewardResult(
    save: save.copyWith(
      gold: save.gold + goldGained,
      bountyClaimedIds: <String>[...save.bountyClaimedIds, bounty.id],
      statistics: PlayerStatistics(
        values: <String, num>{
          ...save.statistics.values,
          'bounties_completed': jsNumber(save.statistics.values['bounties_completed'] ?? 0) + 1,
          'gold_earned': jsNumber(save.statistics.values['gold_earned'] ?? 0) + goldGained,
        },
      ),
    ),
    goldGained: goldGained,
  );
}

File: src/game/bounties/complete.ts
import type { GameDatabase } from '../data/types'
import { getSession, isSignedIn } from '../multiplayer/auth'
import { getLocalBackend } from '../multiplayer/client'
import { inventoryCount } from '../production/recipes'
import type { PlayerSave } from '../save/types'
import { isBountyReadyToClaim, syncBountyHour } from './progress'
import { hourlyBountyBoard } from './rotation'
import type { BountyClaimRecord, BountyDefinition } from './types'

function consumeInventoryItems(
  save: PlayerSave,
  itemId: string,
  amount: number,
): { ok: true; save: PlayerSave } | { ok: false; reason: string } {
  if (inventoryCount(save, itemId) < amount) {
    return { ok: false, reason: 'Not enough items to deliver.' }
  }
  let remaining = amount
  const inventory = save.inventory
    .map((stack) => {
      if (stack.itemId !== itemId || remaining <= 0 || stack.enchantmentId) return stack
      const take = Math.min(stack.quantity, remaining)
      remaining -= take
      return { ...stack, quantity: stack.quantity - take }
    })
    .filter((stack) => stack.quantity > 0)
  if (remaining > 0) return { ok: false, reason: 'Not enough items to deliver.' }
  return { ok: true, save: { ...save, inventory } }
}

export type BountyTurnInCheck =
  | { ok: true; save: PlayerSave }
  | { ok: false; reason: string }

/**
 * Everything the turn-in decides locally: the board is current, the objective is
 * met, and any delivered items have left the bag. The returned save is what the
 * reward is applied to once a backend accepts the claim.
 */
export function prepareBountyTurnIn(
  save: PlayerSave,
  bounty: BountyDefinition,
  nowMs: number,
): BountyTurnInCheck {
  const next = syncBountyHour(save, nowMs)
  const board = hourlyBountyBoard(nowMs)
  if (board.hourKey !== next.bountyHourKey) {
    return { ok: false, reason: 'This bounty board has rotated.' }
  }
  if (!board.bounties.some((row) => row.id === bounty.id)) {
    return { ok: false, reason: 'That bounty is not on the current board.' }
  }
  if ((next.bountyClaimedIds ?? []).includes(bounty.id)) {
    return { ok: false, reason: 'You already claimed this bounty.' }
  }
  if (!isBountyReadyToClaim(next, bounty, nowMs)) {
    return { ok: false, reason: 'Finish the bounty objective first.' }
  }
  if (bounty.kind === 'gather_deliver') {
    return consumeInventoryItems(next, bounty.targetId, bounty.amount)
  }
  return { ok: true, save: next }
}

export interface BountyRewardResult {
  save: PlayerSave
  goldGained: number
}

/**
 * Pay out a bounty a backend has accepted. Everyone eligible earns the base
 * reward; only the claim the server recorded first adds the bonus.
 */
export function applyBountyReward(
  save: PlayerSave,
  bounty: BountyDefinition,
  firstCompleter: boolean,
): BountyRewardResult {
  const goldGained = bounty.rewardGold + (firstCompleter ? bounty.firstPlaceBonusGold : 0)
  return {
    save: {
      ...save,
      gold: save.gold + goldGained,
      bountyClaimedIds: [...(save.bountyClaimedIds ?? []), bounty.id],
      statistics: {
        values: {
          ...save.statistics.values,
          bounties_completed: Number(save.statistics.values.bounties_completed ?? 0) + 1,
          gold_earned: Number(save.statistics.values.gold_earned ?? 0) + goldGained,
        },
      },
    },
    goldGained,
  }
}

export type BountyTurnInResult =
  | {
      ok: true
      save: PlayerSave
      claim: BountyClaimRecord
      firstCompleter: boolean
      goldGained: number
    }
  | { ok: false; reason: string }

/** Plaza notice-board turn-in, against the signed-in player's backend. */
export function turnInBounty(
  _db: GameDatabase,
  save: PlayerSave,
  bounty: BountyDefinition,
  nowMs: number = Date.now(),
): BountyTurnInResult {
  if (!isSignedIn()) return { ok: false, reason: 'Sign in to claim bounties.' }
  const session = getSession()
  if (!session) return { ok: false, reason: 'Sign in to claim bounties.' }

  const prepared = prepareBountyTurnIn(save, bounty, nowMs)
  if (!prepared.ok) return prepared

  const hourKey = hourlyBountyBoard(nowMs).hourKey
  const claimResult = getLocalBackend().claimBounty(session, hourKey, bounty.id)
  if (!claimResult.ok) return { ok: false, reason: claimResult.reason }

  const firstCompleter =
    claimResult.firstCompleter && claimResult.claim.userId === session.userId
  const rewarded = applyBountyReward(prepared.save, bounty, firstCompleter)

  return {
    ok: true,
    save: rewarded.save,
    claim: claimResult.claim,
    firstCompleter,
    goldGained: rewarded.goldGained,
  }
}
